model/scope.go

```diff
@@ -0,0 +1,208 @@
+	return Scope(PluginScopePrefix + pluginID)
+}
+
+func NewPluginSpecificScope(scopeName string) Scope {
+	//TODO this leads to clashes between plugin scopes. Rethink.
+	return Scope(PluginSpecificScopePrefix + scopeName)
+}
+
+func (s Scope) IsPluginScope() bool {
+	return strings.HasPrefix(string(s), PluginScopePrefix)
+}
+
+func (s Scope) IsPluginSpecificScope() bool {
+	return strings.HasPrefix(string(s), PluginSpecificScopePrefix)
+}
+
+func getPredefinedScopes() Scopes {
+	return Scopes{
+		ScopeChannelJoin,
+		ScopeMessageRead,
+		ScopeMessageWrite,
+		ScopeFilesRead,
+		ScopeFilesWrite,
+		ScopeSearchPosts,
+	}
+}
+
+func (s Scope) IsPredefinedScope() bool {
+	return getPredefinedScopes().isInScope(s)
+}
+
+func (s Scope) IsScopeForPlugin(pluginID string) bool {
+	return s.IsPluginScope() &&
+		strings.TrimPrefix(string(s), PluginScopePrefix) == pluginID
+}
+
+// isInScope checks if a scope is in the scope list
+func (ss Scopes) isInScope(scope Scope) bool {
+	if ss == nil {
+		return false
+	}
+
+	for _, allowed := range ss {
+		if scope == allowed {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (ss Scopes) AreAllowed(allowed Scopes) bool {
+	// To support legacy OAuth Apps, we consider nil scopes as a non-scoped OAuth App.
+	if ss == nil {
+		return true
+	}
+
+	// Allowed endpoints will just relay the scopes, in case there is any consideration to be made based on scopes.
+	if allowed.equals(ScopeAllow()) {
+		return true
+	}
+
+	if allowed.equals(ScopeDeny()) {
+		return false
+	}
+
+	if len(ss.intersection(allowed)) == 0 {
+		return false
+	}
+
+	return true
+}
+
+func (ss Scopes) IsPluginInScope(pluginID string) bool {
+	if ss == nil {
+		return false
+	}
+
+	for _, allowed := range ss {
+		if allowed.IsScopeForPlugin(pluginID) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (ss Scopes) intersection(scope Scopes) Scopes {
+	if ss == nil || scope == nil {
+		return ScopeDeny()
+	}
+
+	out := Scopes{}
+	for _, x := range ss {
+		if scope.isInScope(x) {
+			out = append(out, x)
+		}
+	}
+
+	if len(out) == 0 {
+		return ScopeDeny()
+	}
+
+	return out
+}
+
+func (ss Scopes) Validate() bool {
+	for _, s := range ss {
+		if !s.IsPluginScope() &&
+			!s.IsPluginSpecificScope() &&
+			!s.IsPredefinedScope() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (ss Scopes) equals(ss2 Scopes) bool {
+	// TODO OAUTH confirm this is true
+	if ss == nil {
+		return ss2 == nil
+	}
+
+	if ss2 == nil {
+		return false
+	}
+
+	return len(ss) == len(ss2) && len(ss) == len(ss.intersection(ss2))
+}
+
+func (ss Scopes) IsSuperset(ss2 Scopes) bool {
+	if ss == nil {
+		return true
+	}
+
+	if ss2 == nil {
+		return false
+	}
+
+	if len(ss2) > len(ss) {
+		return false
+	}
+
+	for _, s2 := range ss2 {
+		if !ss.isInScope(s2) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Normalize removes all repeated scopes from a list of Scopes.
+func (ss Scopes) Normalize() Scopes {
+	if ss == nil {
+		return nil
+	}
+
+	out := Scopes{}
+
+OUTER:
```

Comment:
There's a cool zero-alloc algorithm I suggest for this use-case: https://github.com/golang/go/wiki/slicetricks#in-place-deduplicate-comparable. 0/5 on whether you want to use it, but if you do, please add a comment explaining what this code is doing.